gpio/src/gpio.rs

```diff
@@ -0,0 +1,943 @@
+                val: None,
+                irq_type: VIRTIO_GPIO_IRQ_TYPE_NONE,
+            }));
+        }
+
+        Ok(GpioController {
+            config: VirtioGpioConfig {
+                ngpio: From::from(ngpio),
+                padding: From::from(0),
+                gpio_names_size: From::from(line_names.len() as u32),
+            },
+            device,
+            line_state,
+            line_names,
+        })
+    }
+
+    pub(crate) fn get_num_lines(&self) -> u16 {
+        self.device.get_num_lines().unwrap()
+    }
+
+    fn get_direction(&self, gpio: u16) -> Result<u8> {
+        Ok(self.line_state[gpio as usize].read().unwrap().dir)
+    }
+
+    fn set_direction(&self, gpio: u16, dir: u32) -> Result<()> {
+        let line_state = &mut self.line_state[gpio as usize].write().unwrap();
+
+        let value = match dir as u8 {
+            VIRTIO_GPIO_DIRECTION_NONE => {
+                line_state.val = None;
+                0
+            }
+
+            VIRTIO_GPIO_DIRECTION_IN => 0,
+            VIRTIO_GPIO_DIRECTION_OUT => match line_state.val {
+                Some(val) => val,
+                None => return Err(Error::GpioCurrentValueInvalid),
+            },
+
+            _ => return Err(Error::GpioDirectionInvalid(dir)),
+        };
+
+        self.device.set_direction(gpio, dir as u8, value as u32)?;
+        line_state.dir = dir as u8;
+        Ok(())
+    }
+
+    fn get_value(&self, gpio: u16) -> Result<u8> {
+        self.device.get_value(gpio)
+    }
+
+    fn set_value(&self, gpio: u16, value: u32) -> Result<()> {
+        if value > 1 {
+            return Err(Error::GpioValueInvalid(value));
+        }
+
+        self.device.set_value(gpio, value)?;
+        self.line_state[gpio as usize].write().unwrap().val = Some(value as u16);
+        Ok(())
+    }
+
+    pub(crate) fn get_irq_type(&self, gpio: u16) -> u16 {
+        self.line_state[gpio as usize].read().unwrap().irq_type
+    }
+
+    fn set_irq_type(&self, gpio: u16, value: u32) -> Result<()> {
+        let irq_type = value as u16;
+
+        // Invalid irq type
+        if (irq_type & !VIRTIO_GPIO_IRQ_TYPE_ALL) != 0 {
+            return Err(Error::GpioIrqTypeInvalid(irq_type));
+        }
+
+        // Begin critical section
+        let line_state = &mut self.line_state[gpio as usize].write().unwrap();
+        let prev_irq_type = line_state.irq_type;
+
+        // New irq type same as current one.
+        if irq_type == prev_irq_type {
+            return Err(Error::GpioIrqTypeNoChange(irq_type));
+        }
+
+        // Must configure to VIRTIO_GPIO_IRQ_TYPE_NONE first before changing irq type.
+        if prev_irq_type != VIRTIO_GPIO_IRQ_TYPE_NONE && irq_type != VIRTIO_GPIO_IRQ_TYPE_NONE {
+            return Err(Error::GpioOldIrqTypeValid(prev_irq_type));
+        }
+
+        self.device.set_irq_type(gpio, irq_type)?;
+        line_state.irq_type = irq_type;
+        Ok(())
+    }
```

Comment:
Same as above - it is hard to understand the motivation behind keeping a shadow state of the GPIO lines when the self.device can be queried.

Reply:
It is an optimization. At least in the current setup, all these calls to hardware go via /dev/gpiochipX node to the kernel. This is expensive for the stuff where we already know what the value is.

Reply:
The problem with keeping shadow data structures that supposedly reflect the state of the HW is that they always get out of sync, creating bugs that are very difficult to track down.  I also wonder if optimization is a real concern?  How often do GPIOs get reconfigured in typical scenarios?

I would personally get rid of the shadow data structure but other people may disagree.  If we end up keeping it then some in-line documentation is needed to explain what is happening and the motivations for doing so.

Reply:
I have update get_direction() to go fetch the direction again now.

The shadow state though needs to stay as we need to check direction, value, irq_type at various places while handling new requests.

Reply:
I think it's fair enough if the backend is the only thing that will every change things (e.g. has exclusive access to the GPIO). Personally I'd call it a cache rather than a shadow value but that's just personal preference. You could have a debug build verify the cache value with the actual device state each time it is accessed.

Reply:
> I think it's fair enough if the backend is the only thing that will every change things

That's what I thought. But then it was just a single line fetching direction from hardware, so I let it go :)